### ISSUE TITLE
Any way to create a complex `style` or a responsive `variant`?

### ISSUE THREAD

**Issue by talbet** [jxnblk/styled-system#314]
My use case is as follows: I need to make a form label 'visually hidden' on some screens, and visible on others. For accessibility this cannot be a simple `display: none` and it needs to apply a several styles.

Using custom `style` declarations, this might be possible, although it would not be neat as it would require binding a bunch of different properties to the same prop.

Using custom `variant` keys there doesn't seem to be a way to access breakpoints.

Is there some combination of the two that would work? Ideally it would look something like:

```
<Label visuallyHidden={[true, false, false]}>My form label</Label>
```

**Comment by jxnblk** [comment]
Currently there isn't a notion of a responsive variant – this seems like a fairly small use-case, and I'd recommend adding your own prop to handle this sort of thing

**Comment by Jrayp** [comment]
I am also trying to do something like a responsive variant. Without any kind of responsiveness, I can't find too much use for variants on their own (though I may be missing something).

It would be great if I could do something like:
```js
const headings = variant({ key: "headings" });
export const Heading = styled(Text)(themed("Heading"), headings);
```
And inside of the theme object:
```js
	headings: {
		small: {
			fontSize: ["1em", "1.25em", "1.5em" ]
		},
		big: {
			fontSize: ["2em", "2.5em", "3em" ]
		}
	}
```

Based on your response here, and to some other similar issues don't see this happening (and it may not make sense, or just not be very viable), but how do you recommend we do something like this cleanly?

Right now I just use a function component, and literally pull the values from a map, but this is pretty messy imo.

**Comment by jxnblk** [comment]
@Jrayp my suggestion for your example is to use props (e.g. map props) and avoid mixing props objects with the theme & style objects

**Comment by avidrucker** [comment]
> @Jrayp my suggestion for your example is to use props (e.g. map props) and avoid mixing props objects with the theme & style objects

@jxnblk Do you have any examples that show clear delineation of theme, style, and props objects in context? Why one would use one and not the other in typical use cases? I am looking at https://jxnblk.com/styled-system/ and trying to find something akin to (A) https://tachyons.io/components/ (usage in isolation/vaccum) and (B) https://tachyons.io/gallery/ (usage in context/actual environment) where I'm struggling to find some B's. Perhaps I am missing a good link/reference... Please advise!

**Comment by jxnblk** [comment]
There is a built-in `buttonStyle` function, so you shouldn't need to define your own

**Comment by fitiskin** [comment]
@jxnblk `buttonStyle` isn't responsive, is it?

**Comment by jxnblk** [comment]
Oh, sorry I mixed this issue up with a different one, carry on

**Comment by kibs** [comment]
Hi, I am having the same requirements as @Jrayp, in the design I am trying to implement. There are a limited set of text styles from the designer, but all textStyles, changes in size when going from mobile to desktop.

The same is going on with buttons and other elements. Eg. the "primary" variant of a button, adds more padding when going from mobile to desktop.

**Comment by jxnblk** [comment]
Generally I think some of this is outside of the scope that `variant` was intended to handle, so I'm going to close this issue out.

As another suggestion for how to handle responsive variants, media queries *can* be used in the theme's style objects for variants, for when you want things to change responsively across the board

**Comment by Daveawb** [comment]
> Currently there isn't a notion of a responsive variant – this seems like a fairly small use-case, and I'd recommend adding your own prop to handle this sort of thing

The Design systems that I've been subjected to recently have typically defined different style rules at varying breakpoints for elements such as `<p>`, `<h1>` and so on. Not having a utility to define this in a theme reduces its re-usability across projects considerably.

I'm not saying whether this is right or wrong for DS' to define this, however not being able to do it in the theme makes me sad.

**Comment by jxnblk** [comment]
@Daveawb Sorry, I don't mean to make you sad :/ I'm not opposed to making the variant utility work in some responsive way, but the general convention in Styled System is that responsiveness is a concern of the component, not the theme. That said, I think this WIP PR is a good indication of intent to make variants more powerful, FWIW #583

**Comment by Daveawb** [comment]
@jxnblk Aww I didn't mean it like that, just a throw away comment. I'm actually seeing this now in many many many design systems. I'm still unsure whether it's an anti pattern for DS' where designers and UX teams alike still don't really understand the value of it. I suspect it's a bit of both, either way thanks for the link, this made a massive difference.

**Comment by jxnblk** [comment]
@Daveawb since this issue is pretty old and for clarity, [*variants*](https://styled-system.com/variants) now support responsive values, e.g.:

```js
variant({
  variants: {
    big: {
      fontSize: [ 3, 4, 5 ],
    },
    small: {
      fontSize: [ 1, 2, 3 ],
    }
  }
})
```

**Comment by kjbrum** [comment]
@jxnblk, is something like the following possible, or would it be possible to add the functionality? Similar to what we are able to do with components. https://styled-system.com/responsive-styles#using-objects

```
variant({
  variants: {
    big: {
      fontSize: {
        _: 3,
        md: 4,
        xl: 5
      },
    },
    small: {
      fontSize: {
        _: 1,
        md: 2,
        xl: 3
      },
    }
  }
})
```
